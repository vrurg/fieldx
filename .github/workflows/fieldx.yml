name: CI

on:
  push:
  pull_request:

env:
  CARGO_TERM_COLOR: always
  CARGO_INCREMENTAL: 0

jobs:
  with_features:
    runs-on: ${{ matrix.os }}-latest
    strategy:
      matrix:
        os: [ubuntu, macos, windows]
        toolchain:
          - 1.77
          - stable
          - beta
          - nightly
        feature: [serde]
    steps:
      - uses: actions/checkout@v4
      - run: rustup update ${{ matrix.toolchain }} && rustup default ${{ matrix.toolchain }}
      - run: cargo build --verbose --all --features ${{ matrix.feature }}
      - run: cargo test --verbose --all --features ${{ matrix.feature }}
      - run: cargo test --verbose --all --features diagnostics
        if: matrix.toolchain == 'nightly'
      - run: cargo test --verbose --all --all-features
        if: matrix.toolchain == 'nightly'
  no_features:
    runs-on: ${{ matrix.os }}-latest
    strategy:
      matrix:
        os: [ubuntu, macos, windows]
        toolchain:
          - 1.77
          - stable
          - beta
          - nightly
    steps:
      - uses: actions/checkout@v4
      - run: rustup update ${{ matrix.toolchain }} && rustup default ${{ matrix.toolchain }}
      - run: cargo build --verbose --all
      - run: cargo test --verbose --all
  build_docs:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - run: rustup default nightly
      - run: cargo install cargo-docs-rs
      - run: cargo docs-rs -p fieldx

