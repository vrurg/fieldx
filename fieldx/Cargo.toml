[package]
    authors.workspace      = true
    categories.workspace   = true
    description            = "Procedural macro for constructing structs with lazily initialized fields, builder pattern, and serde support with a focus on declarative syntax."
    edition.workspace      = true
    exclude.workspace      = true
    keywords.workspace     = true
    license-file.workspace = true
    name                   = "fieldx"
    readme                 = "../README.md"
    repository.workspace   = true
    version                = "0.1.8-beta.2"

    # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
    fieldx_aux    = { workspace = true }
    fieldx_derive = { workspace = true, features = ["tracing"] }
    getset        = { workspace = true }
    parking_lot   = { workspace = true }
    serde         = { workspace = true, optional = true }
    thiserror     = { workspace = true }
    tokio         = { workspace = true, features = ["sync"] }

[dev-dependencies]
    cargo-toolchain = { workspace = true }
    crossbeam       = { workspace = true }
    num_cpus        = { workspace = true }
    parking_lot     = { workspace = true }
    rustc_version   = { workspace = true }
    serde_json      = { workspace = true }
    tokio           = { workspace = true, default-features = false, features = ["macros", "rt", "rt-multi-thread", "time"] }
    trybuild        = { workspace = true }

[features]
    diagnostics = ["fieldx_derive/diagnostics"]
    send_guard  = ["parking_lot/send_guard"]
    serde       = ["fieldx_derive/serde", "parking_lot/serde", "dep:serde"]
