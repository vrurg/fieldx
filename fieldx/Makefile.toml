[tasks.test-feature-sets]
    category = "Test"
    clear = true
    description = "Test different feature sets"
    script = '''
!include_files ../makefile.ds

echo Testing feature sets with toolchain '${__FIELDX_DEFAULT_TOOLCHAIN__}'

features = array sync serde
feature_sets = generate_power_set ${features}

for fs in ${feature_sets}
    fs = array_join ${fs} ","
    echo "Testing feature set '${fs}'"
    set_env __FIELDX_FEATURE_SET__ ${fs}
    cm_run_task test-with-features
    cm_run_task test-with-features-async
end
'''
    script_runner = "@duckscript"
    toolchain = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.test-with-features-async]
    run_task = { name = ["test-async-lock", "test-async-tokio"], fork = true }

[tasks.test-async-lock]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-lock" }
    run_task = "test-with-features"

[tasks.test-async-tokio]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-tokio" }
    run_task = "test-with-features"

[tasks.test-with-features]
    category    = "Test"
    description = "Test code compilability with '${__FIELDX_FEATURE_SET__}' features"
    script_runner = "@duckscript"
    script = '''
exec --fail-on-error cargo +${__FIELDX_DEFAULT_TOOLCHAIN__} test --features ${__FIELDX_FEATURE_SET__}
exec --fail-on-error cargo +${__FIELDX_DEFAULT_TOOLCHAIN__} test --features ${__FIELDX_FEATURE_SET__} --examples
'''

# Compilation is useful for updating the error outputs per version+features
[tasks.test-compilation]
    category = "Test"
    clear = true
    description = "Test code compilability"
    script = '''
!include_files ../makefile.ds

echo Testing compilability with toolchain '${__FIELDX_DEFAULT_TOOLCHAIN__}'
cm_run_task compile-default-features

features = array sync serde
feature_sets = generate_power_set ${features}

for fs in ${feature_sets}
    fs = array_join ${fs} ","
    echo "Testing feature set '${fs}' on ${__FIELDX_DEFAULT_TOOLCHAIN__}"
    set_env __FIELDX_FEATURE_SET__ ${fs}
    cm_run_task compile-with-features
    cm_run_task compile-with-features-async
end
'''
    script_runner = "@duckscript"
    toolchain = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.compile-default-features]
    condition   = { env_not = { __FIELDX_DEFAULT_TOOLCHAIN__ = "nightly" } }
    args        = ["test", "--test", "compilation", "--", "--nocapture"]
    category    = "Test"
    command     = "cargo"
    description = "Test code compilability with default feature set"
    toolchain   = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.compile-with-features]
    condition   = { env_not = { __FIELDX_DEFAULT_TOOLCHAIN__ = "nightly" } }
    args        = ["test", "--test", "compilation", "--features", "${__FIELDX_FEATURE_SET__}", "--", "--nocapture"]
    category    = "Test"
    command     = "cargo"
    description = "Test code compilability with '${__FIELDX_FEATURE_SET__}' features"
    toolchain   = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.compile-with-features-async]
    run_task = { name = ["compile-async-lock", "compile-async-tokio"], fork = true }

[tasks.compile-async-lock]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-lock" }
    run_task = "compile-with-features"

[tasks.compile-async-tokio]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-tokio" }
    run_task = "compile-with-features"
