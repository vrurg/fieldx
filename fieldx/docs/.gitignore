.markdownlint.json
doctest_cache
target-doc/
book/