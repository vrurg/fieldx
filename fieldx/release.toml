pre-release-hook = ['/bin/sh', '-c', '''
if [ "$CARGO_MAKE_TASK" != "publish" -a "$CARGO_MAKE_TASK" != "beta" ]; then
    echo "***ERROR*** Must be released with help of 'cargo make publish' in project workspace dir." 1>&2
    exit 1
else
    pushd $WORKSPACE_ROOT >/dev/null
    export CARGO_RELEASE_VERSION={{version}}
    if ! makers changelog; then
        echo "***ERROR*** Changelog generation failed." 1>&2
        exit 1
    fi
    popd
fi
''']

pre-release-replacements = [
    { file = "src/lib.rs", search = "(doc\\s*\\(html_root_url\\s*=\\s*\"https://docs.rs/.+?)/\\d+\\.\\d+\\.\\d.*/\"", replace = "$1/{{version}}/\"", prerelease = true },
]

enable-all-features = false
enable-features     = ["sync", "async-tokio", "serde", "clonable-lock"]
