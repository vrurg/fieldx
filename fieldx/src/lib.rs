pub mod errors;
pub mod traits;

pub use fieldx_derive::fxstruct;
pub use parking_lot::{MappedRwLockReadGuard, RwLock, RwLockReadGuard, RwLockUpgradableReadGuard, RwLockWriteGuard};
use std::{any, cell::RefCell, marker::PhantomData, sync::atomic::AtomicBool};
pub use std::{cell::OnceCell, fmt, sync::atomic::Ordering};
use traits::FXStructSync;

pub struct FXProxy<O, T>
where
    O: FXStructSync,
{
    value: RwLock<Option<T>>,
    is_set: AtomicBool,
    builder: RwLock<Option<fn(&O) -> T>>,
}

#[allow(private_bounds)]
pub struct FXWrLock<'a, O, T, FX>
where
    O: FXStructSync,
    FX: FXProxyCore<O, T>,
{
    lock: RefCell<RwLockWriteGuard<'a, Option<T>>>,
    fxproxy: &'a FX,
    _phantom: PhantomData<O>,
}

impl<O, T> Default for FXProxy<O, T>
where
    O: FXStructSync,
{
    fn default() -> Self {
        Self {
            value: RwLock::new(None),
            is_set: AtomicBool::new(false),
            builder: RwLock::new(None),
        }
    }
}

impl<O, T: fmt::Debug> fmt::Debug for FXProxy<O, T>
where
    O: FXStructSync,
{
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let vlock = self.value.read();
        formatter
            .debug_struct(any::type_name::<Self>())
            .field("value", &*vlock)
            .finish()
    }
}

impl<O, T> From<T> for FXProxy<O, T>
where
    O: FXStructSync,
    Self: FXProxyCore<O, T>,
{
    fn from(value: T) -> Self {
        Self {
            value: RwLock::new(Some(value)),
            is_set: AtomicBool::new(true),
            builder: RwLock::new(None),
        }
    }
}

pub trait FXProxyCore<O, T>
where
    O: FXStructSync,
    Self: Sized,
{
    fn builder(&self) -> &RwLock<Option<fn(&O) -> T>>;
    fn value(&self) -> &RwLock<Option<T>>;
    fn is_set_raw(&self) -> &AtomicBool;

    fn is_set(&self) -> bool {
        self.is_set_raw().load(Ordering::SeqCst)
    }

    fn proxy_setup(&self, builder: fn(&O) -> T) {
        *self.builder().write() = Some(builder);
        if !self.is_set() && (*self.value().read()).is_some() {
            self.is_set_raw().store(true, Ordering::SeqCst);
        }
    }

    fn read<'a>(&'a self, owner: &O) -> MappedRwLockReadGuard<'a, T> {
        let guard = self.value().upgradable_read();
        if (*guard).is_none() {
            let mut wguard = RwLockUpgradableReadGuard::upgrade(guard);
            // Still uninitialized? Means no other thread took care of it yet.
            if wguard.is_none() {
                // No value has been set yet
                match *self.builder().read() {
                    Some(ref builder_cb) => {
                        *wguard = Some((*builder_cb)(owner));
                        self.is_set_raw().store(true, Ordering::SeqCst);
                    }
                    None => panic!("Builder is not set"),
                }
            }
            return RwLockReadGuard::map(RwLockWriteGuard::downgrade(wguard), |data: &Option<T>| {
                data.as_ref().unwrap()
            });
        }
        RwLockReadGuard::map(RwLockUpgradableReadGuard::downgrade(guard), |data: &Option<T>| {
            data.as_ref().unwrap()
        })
    }

    fn write<'a>(&'a self) -> FXWrLock<'a, O, T, Self> {
        FXWrLock::<'a, O, T, Self>::new(self.value().write(), self)
    }

    fn clear_with_lock(&self, wguard: &mut RwLockWriteGuard<Option<T>>) -> Option<T> {
        self.is_set_raw().store(false, Ordering::SeqCst);
        wguard.take()
    }

    fn clear(&self) -> Option<T> {
        let mut wguard = self.value().write();
        self.clear_with_lock(&mut wguard)
    }
}

impl<O, T> FXProxyCore<O, T> for FXProxy<O, T>
where
    O: FXStructSync,
{
    #[inline]
    fn builder(&self) -> &RwLock<Option<fn(&O) -> T>> {
        &self.builder
    }

    #[inline]
    fn value(&self) -> &RwLock<Option<T>> {
        &self.value
    }

    #[inline]
    fn is_set_raw(&self) -> &AtomicBool {
        &self.is_set
    }
}

impl<O, T> FXProxy<O, T> where O: FXStructSync {}

#[allow(private_bounds)]
impl<'a, O, T, FX> FXWrLock<'a, O, T, FX>
where
    O: FXStructSync,
    FX: FXProxyCore<O, T>,
{
    pub fn new(lock: RwLockWriteGuard<'a, Option<T>>, fxproxy: &'a FX) -> Self
    where
        FX: FXProxyCore<O, T>,
    {
        let lock = RefCell::new(lock);
        Self {
            lock,
            fxproxy,
            _phantom: PhantomData,
        }
    }

    pub fn store(&mut self, value: T) -> Option<T> {
        self.fxproxy.is_set_raw().store(true, Ordering::Release);
        self.lock.borrow_mut().replace(value)
    }

    pub fn clear(&self) -> Option<T> {
        self.fxproxy.clear_with_lock(&mut *self.lock.borrow_mut())
    }
}
