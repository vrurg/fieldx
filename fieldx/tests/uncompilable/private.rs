mod inner {
    use fieldx::fxstruct;

    #[fxstruct(sync)]
    pub struct Inner {
        #[fieldx(vis(), lazy, reader)]
        foo: u8,
    }

    impl Inner {
        fn build_foo(&self) -> u8 {
            42
        }
    }
}

fn main() {
    let inner = inner::Inner::new();
    // This must not compile, as expected.
    let foo = inner.read_foo();
}
