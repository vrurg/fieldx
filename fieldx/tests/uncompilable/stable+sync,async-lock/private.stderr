error[E0624]: method `read_foo` is private
 --> tests/uncompilable/private.rs
  |
  |         #[fieldx(vis(), lazy, reader)]
  |                               ------ private method defined here
...
  |     let foo = inner.read_foo();
  |                     ^^^^^^^^ private method
